### ISSUE TITLE
Material#update is very slow

### ISSUE THREAD

**Issue by willeastcott** [playcanvas/engine#699]
Calling material update is extremely slow and inefficient. For example, check out this UV scrolling function:

```javascript
Uvscroll.prototype.update = function(dt) {
    var t = (new Date().getTime() / 2000) % 1;
    var material = this.entity.model.material;
    material.diffuseMapOffset.set(t, t);
    material.update();
};

```

The call to update factors highly in a CPU profile of an app.

Also note that many developers simply forget to call it, so a refactor which render update redundant would be ideal.

**Comment by guycalledfrank** [comment]
You can scroll textures without using the update function: https://playcanvas.com/editor/scene/372898
But seems like nobody uses it...

**Comment by willeastcott** [comment]
No, I don't even think that's officially API (and I gotta be honest, that code to scroll the UVs is pretty ugly!). My snippet is the official way. Removing the update stops the scrolling effect.

**Comment by willeastcott** [comment]
So I've been thinking.... 😄 

Perhaps we should treat materials like textures. Whenever you update a property, we just set a dirty flag to true. Then, in the forward renderer, when the material is set, we check the flag and update the material internally as required.

And the update procedure should just update what's changed. Maybe we need some kind of 'dirty properties' list.

**Comment by guycalledfrank** [comment]
I should definitely refactor this mess

**Comment by daredevildave** [comment]
Fixed in #1373
